python_modules/dagster/dagster/core/execution/execute_job_result.py

```diff
@@ -0,0 +1,118 @@
+from typing import Any, Sequence
+
+import dagster._check as check
+from dagster.core.definitions import JobDefinition, NodeHandle
+from dagster.core.errors import DagsterInvariantViolationError
+from dagster.core.events import DagsterEvent
+from dagster.core.execution.plan.step import StepKind
+from dagster.core.execution.plan.utils import build_resources_for_manager
+from dagster.core.storage.pipeline_run import DagsterRun
+
+from .execution_result import ExecutionResult
+
+
+class ExecuteJobResultContext:
+    def __init__(self, job_def, reconstruct_context, event_list, dagster_run):
+        self._job_def = job_def
+        self._reconstruct_context = reconstruct_context
+        self._event_list = event_list
+        self._dagster_run = dagster_run
+
+
+class ExecuteJobResult(ExecutionResult):
+    def __init__(self, job_def, reconstruct_context, event_list, dagster_run):
+        self._job_def = job_def
+        self._reconstruct_context = reconstruct_context
+        self._context = None
+        self._event_list = event_list
+        self._dagster_run = dagster_run
+
+    def __enter__(self) -> "ExecuteJobResult":
+        context = self._reconstruct_context.__enter__()
+        self._context = context
+        return self
+
+    def __exit__(self, *exc):
+        return self._reconstruct_context.__exit__(*exc)
+
+    @property
+    def job_def(self) -> JobDefinition:
+        return self._job_def
+
+    @property
+    def dagster_run(self) -> DagsterRun:
+        return self._dagster_run
+
+    @property
+    def all_events(self) -> Sequence[DagsterEvent]:
+        return self._event_list
+
+    @property
+    def run_id(self) -> str:
+        return self.dagster_run.run_id
+
+    def compute_events_for_handle(self, handle: NodeHandle) -> Sequence[DagsterEvent]:
+        return [
+            event
+            for event in self._filter_events_by_handle(handle)
+            if event.step_kind == StepKind.COMPUTE
+        ]
+
+    def _get_output_for_handle(self, handle: NodeHandle, output_name: str) -> Any:
+        if not self._context:
+            raise DagsterInvariantViolationError(
+                "In order to access output objects, the result of `execute_job` must be opened as a context manager: 'with execute_job(...) as result:"
+            )
```

Comment:
This pattern is a little tricksy, where we have an optionally context manger object and some methods throw if you didn't enter. The alternative is introducing an different object that you have to context manager enter to get that has the result methods. Trade off from my view is api surface area vs static analyzability.

I'm fine with either approach, just want to make sure we make this trade off decision explicitly.

Reply:
The main benefit of the tighter surface area appears to be `ExecutionResult` base where output access is common between `execute_job` and `execute_in_process`. That is nice, though does sort of hide the fundamental change in behavior underlying the output fetching between the two approaches. I think `ExecutionResult` not having output fetching as common methods and having the divergent behavior modeled as different apis has its merits.

Reply:
I feel like if communicated effectively in docs, then it doesn't make a huge difference. I've generally found that when there are two separate feature stacks that are parallel, there will be feature slippage, and I think having a shared interface will hopefully help to prevent that.

Reply:
> communicated effectively in docs

[1]

are there going to be additional docs? should those just be in this PR too?

Reply:
will add